app/src/common/shared/org/mozilla/vrbrowser/History.java

```diff
@@ -0,0 +1,210 @@
+            }
+            try {
+                JSONArray array = new JSONArray(sb.toString());
+                mItems.clear();
+                for (int i = 0; i < array.length(); ++i) {
+                    Item item = new Item(array.getJSONObject(i));
+                    if (item.url != null && item.url.length() > 0) {
+                        mItems.put(key(item.url), item);
+                    }
+                }
+
+            } catch (JSONException e) {
+                Log.e(LOGTAG, "Error parsing history", e);
+            }
+        }
+        catch (Exception ex) {
+            Log.e(LOGTAG, "Error reading history", ex);
+        }
+    }
+
+    private String getJSONString() {
+        JSONArray array = new JSONArray();
+        for (Item item: mItems.values()) {
+            array.put(item.toJSON());
+        }
+        return array.toString();
+    }
+
+    private String key(String aUri) {
+        String result = aUri.toLowerCase();
+        int index = result.indexOf('#');
+        if (index >= 0) {
+            result = result.substring(0, index);
+        }
+        return result;
+    }
+
+    private void checkMaxSize() {
+        if (mItems.size() <= MAX_ITEMS) {
+            return;
+        }
+
+        ArrayList<Item> list = new ArrayList<>(mItems.values());
+        list.sort(new Comparator<Item>() {
+            @Override
+            public int compare(Item o1, Item o2) {
+                return (int)(o1.timestamp - o2.timestamp);
+            }
+        });
+
+        for (Item item: list) {
+            mItems.remove(key(item.url));
+            if (mItems.size() >= MAX_ITEMS) {
+                return;
+            }
+        }
+    }
+
+    // GeckoSession.ProgressDelegate
+
+    @Override
+    public void onPageStart(GeckoSession geckoSession, String s) {
+
+    }
+
+    @Override
+    public void onPageStop(GeckoSession geckoSession, boolean b) {
+        String uri = SessionStore.get().getUriFromSession(geckoSession);
+        if (uri == null || uri.length() == 0) {
+            return;
+        }
+        if (uri.toLowerCase().startsWith(SessionStore.HOME_WITHOUT_REGION_ORIGIN)) {
+            return;
+        }
+        if (geckoSession.getSettings().getBoolean(GeckoSessionSettings.USE_PRIVATE_MODE)) {
+            // Do not store private mode history
+            return;
+        }
+
+        Item item = mItems.get(key(uri));
+        if (item == null) {
+            item = new Item();
+            item.url = uri;
+            mItems.put(key(uri), item);
+            checkMaxSize();
+        }
+
+        item.url = uri;
+        item.title = SessionStore.get().getTitleFromSession(geckoSession);
+        item.timestamp = System.currentTimeMillis();
+        save();
+    }
+
+    @Override
+    public void onProgressChange(GeckoSession geckoSession, int i) {
+
+    }
+
+    @Override
+    public void onSecurityChange(GeckoSession geckoSession, SecurityInformation securityInformation) {
+
```

Comment:
it'd probably be wise to check `isSecure`, in case https://webxr.today gets MITM'd or the cert expires or something - in that case, we should not pass the hash to the page

Reply:
I handled this in commit 3d7272b where the Home URI gets set in `SessionStore`: https://github.com/MozillaReality/FirefoxReality/pull/457/files#diff-70a8baac173fb5a93995a46725b70c73R440